controllers/argocd/applicationset/role.go

```diff
@@ -0,0 +1,166 @@
+		ObjectMeta: metav1.ObjectMeta{
+			Name:        resourceName,
+			Namespace:   asr.Instance.Namespace,
+			Labels:      resourceLabels,
+			Annotations: asr.Instance.Annotations,
+		},
+		Rules:     getPolicyRules(),
+		Client:    asr.Client,
+		Mutations: []mutation.MutateFunc{mutation.ApplyReconcilerMutation},
+	}
+
+	desiredRole, err := permissions.RequestRole(roleRequest)
+	if err != nil {
+		asr.Logger.Error(err, "reconcileRole: failed to request role", "name", desiredRole.Name, "namespace", desiredRole.Namespace)
+		asr.Logger.V(1).Info("reconcileRole: one or more mutations could not be applied")
+		return err
+	}
+
+	namespace, err := cluster.GetNamespace(asr.Instance.Namespace, asr.Client)
+	if err != nil {
+		asr.Logger.Error(err, "reconcileRole: failed to retrieve namespace", "name", asr.Instance.Namespace)
+		return err
+	}
+	if namespace.DeletionTimestamp != nil {
+		if err := asr.DeleteRole(desiredRole.Name, desiredRole.Namespace); err != nil {
+			asr.Logger.Error(err, "reconcileRole: failed to delete role", "name", desiredRole.Name, "namespace", desiredRole.Namespace)
+		}
+		return err
+	}
+
+	existingRole, err := permissions.GetRole(desiredRole.Name, desiredRole.Namespace, asr.Client)
+	if err != nil {
+		if !errors.IsNotFound(err) {
+			asr.Logger.Error(err, "reconcileRole: failed to retrieve role", "name", existingRole.Name, "namespace", existingRole.Namespace)
+			return err
+		}
+
+		if err = controllerutil.SetControllerReference(asr.Instance, desiredRole, asr.Scheme); err != nil {
+			asr.Logger.Error(err, "reconcileRole: failed to set owner reference for role", "name", desiredRole.Name, "namespace", desiredRole.Namespace)
+		}
+
+		if err = permissions.CreateRole(desiredRole, asr.Client); err != nil {
+			asr.Logger.Error(err, "reconcileRole: failed to create role", "name", desiredRole.Name, "namespace", desiredRole.Namespace)
+			return err
+		}
+		asr.Logger.V(0).Info("reconcileRole: role created", "name", desiredRole.Name, "namespace", desiredRole.Namespace)
+		return nil
+	}
+
+	if !reflect.DeepEqual(existingRole.Rules, desiredRole.Rules) {
+		existingRole.Rules = desiredRole.Rules
+		if err = permissions.UpdateRole(existingRole, asr.Client); err != nil {
+			asr.Logger.Error(err, "reconcileRole: failed to update role", "name", existingRole.Name, "namespace", existingRole.Namespace)
+			return err
+		}
+	}
+	asr.Logger.V(0).Info("reconcileRole: role updated", "name", existingRole.Name, "namespace", existingRole.Namespace)
+	return nil
+}
+
+func (asr *ApplicationSetReconciler) DeleteRole(name, namespace string) error {
+	if err := permissions.DeleteRole(name, namespace, asr.Client); err != nil {
+		asr.Logger.Error(err, "DeleteRole: failed to delete role", "name", name, "namespace", namespace)
+		return err
+	}
+	asr.Logger.V(0).Info("DeleteRole: role deleted", "name", name, "namespace", namespace)
+	return nil
+}
+
+func getPolicyRules() []rbacv1.PolicyRule {
+	return []rbacv1.PolicyRule{
```

Comment:
@ciiay can we update these permissions to match https://github.com/argoproj/argo-cd/blob/master/manifests/install.yaml#L20351?

Reply:
Sure. Then this means the current roles in applicationset controller doesn't match the manifests you provided, and it was okay?

Reply:
yeah its fine
maybe we just didnt update them
this seems like a good chance to get aligned again

Reply:
Cool, as long as you are sure the manifest ones are more up to date and reliable.